From 4e3fbc9abd29c1bb87b5f197c238661dc653fdf8 Mon Sep 17 00:00:00 2001
From: Mustapha Ben Zoubeir <mustaphax.ben.zoubeir@intel.com>
Date: Tue, 23 Sep 2014 11:17:54 +0200
Subject: [PATCH 4/8] n_gsm:check the dlci pointer in chars_in_buffer

In case of line discipline detach while a mux client has
received a POLL_IN event, the chars_in_buffer function
now check the dlci pointer value to avoid kernel crash.

Change-Id: Ia60cb0b1b5e4eaae01a7bccada6a7a86f2d05227
orig-Change-Id: I9ddd8769b2bc688196e4501472f80521030227d7
Signed-off-by: Gwenn Bourree <gwenn.bourree@intel.com>
Signed-off-by: Nicolas LOUIS <nicolasx.louis@intel.com>
Reviewed-on: http://android.intel.com:8080/128582
Reviewed-by: Ravindran, Arun <arun.ravindran@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 drivers/tty/n_gsm.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/tty/n_gsm.c b/drivers/tty/n_gsm.c
index 43d2574b1566..497a775f9ac9 100644
--- a/drivers/tty/n_gsm.c
+++ b/drivers/tty/n_gsm.c
@@ -3315,9 +3315,16 @@ static int gsmtty_write_room(struct tty_struct *tty)
 
 static int gsmtty_chars_in_buffer(struct tty_struct *tty)
 {
-	struct gsm_dlci *dlci = tty->driver_data;
-	if (dlci->state == DLCI_CLOSED)
+	struct gsm_dlci *dlci;
+
+	if (!tty)
+		return 0;
+
+	dlci = tty->driver_data;
+
+	if ((!dlci) || (dlci->state == DLCI_CLOSED))
 		return -EINVAL;
+
 	return kfifo_len(dlci->fifo);
 }
 
-- 
1.9.1

